### ISSUE TITLE
possible to force saving the terrain?

### ISSUE THREAD

**Issue by nonchip** [Zylann/godot_voxel#160]
i want to allow the user to save the game at any point without quitting or moving far enough away from the blocks to save to immerge them, is there a way to trigger saving a terrain or block? looking at `VoxelTerrain` all the functions that deal with this kinda thing seem private.

**Comment by menip** [comment]
This seems to work for me: https://github.com/Zylann/godot_voxel/issues/146#issuecomment-663304673

**Comment by Zylann** [comment]
There is a private function `save_all_modified_blocks` doing that but it can't be exposed as is (in particular its argument when set to false, which can be dangerous is not used properly).
It iterates all blocks to find the modified ones, so it might cause a stall.

**Comment by nonchip** [comment]
> It iterates all blocks to find the modified ones, so it might cause a stall.

that's not too bad when the user actively clicks save i guess, but yeah might be something to be aware of.

> Note: saving is asynchronous, so it won't be complete after the function is called.

hmm maybe then when exposing that function there should be some kind of callback or state that one can watch to see whether it's done?

> and is what makes autosave fail atm. Is this why you want the feature?

i just want a "Save" button in my menu without having to quit the game/scene.

oh also if there was a `save_this_one_block` function that could maybe be used together with `immerge_block` to implement networking, but that's not my primary goal here, might be useful to keep in mind though.

**Comment by menip** [comment]
Additionally, this feature would allow for snapshots of terrain.

**Comment by Zylann** [comment]
> hmm maybe then when exposing that function there should be some kind of callback or state that one can watch to see whether it's done?

That's a bit more complicated to do this at the moment, so I'm not sure it's worth doing it for now. One way to check this is to remember how many blocks were sent for saving, have the thread return acknowledgements and compare the counts, which provides a progress.

> Additionally, this feature would allow for snapshots of terrain.

It won't, it will just save the unsaved parts from the part of the world currently in memory.
If you want to force saving a section of the loaded world, you'd need something else (which you can do already in GDScript I believe).

**Comment by menip** [comment]
> Not really, it will just save the unsaved blocks from the part of the world currently in memory, in the save directory.
> If you want to extract a section of the loaded world, you'd need something else (which you can do already in GDScript I believe, by litterally using `get_voxel` on the area you want).

Have current save. Modifications are made to terrain. Then when player clicks snapshot button, a copy of save directory is made and set as new directory. Modifications are then written to this new directory. Then at a later time, player can select which terrain snapshot they want to load. This can of course be made more efficient, but it's for example of functionality.

**Comment by Zylann** [comment]
Yeah, though doing something like that implies serious change because that assumes everything is synchronous, while it's not.

**Comment by nonchip** [comment]
> One way to check this is to remember how many blocks were sent for saving, have the thread return acknowledgements and compare the counts, which provides a progress you can check.

that would be totally fine for those purposes. i don't want it to be synchronous, the multithreaded approach is great for performance, just being able to do "hey save that block please" and "have you saved that block yet" would be totally fine.

but being able to save the game when it (and by extension the user) wants instead of when a single component feels like it's time sounds like a pretty important feature :D

**Comment by Zylann** [comment]
I just added two methods to trigger save of one, or all modified blocks in ec6bc00562fe6f613198493a4c0b4f623272e83c

**Comment by menip** [comment]
@Zylann When saving single block, why do we not copy? https://github.com/Zylann/godot_voxel/blob/ec6bc00562fe6f613198493a4c0b4f623272e83c/terrain/voxel_terrain.cpp#L1093-L1106

**Comment by Zylann** [comment]
Oh, it actually should indeed. I copied it from `immerge_block`. Going to fix it in a sec
